File: Lang-K/src/lang/token/TokenSet.java
package lang.token;


public class TokenSet {
	private final Token[] list;
	private final int length;
	private int offset = 0;
	
	public TokenSet(Token[] tokens) {
		this.list = tokens;
		length = tokens.length;
	}
	
	public void unget() {
		if (offset > 0) offset--;
	}
	public void unget(int k) {
		if (offset >= k) offset -= k;
	}
	public Token next() {
		return list[offset++];
	}
	public Token get() {
		return list[offset];
	}
	
	public boolean isEOF() {
		return offset >= length;
	}
	
	public boolean isMatch(String... args) {
		String t = get().string;
		for (String str: args) {
			if (str.equals(t)) return true;
		}
		return false;
	}
	public boolean read(String... args) {
		for (String t : args) {
	 		Token next = next();
	 		if (!t.equals(next.string)) return false;
		}
		return true;
	}
	
	public boolean isName() {
		return isEOF()? false : get().kind == TokenKind.Identifier;
	}
	
	@Deprecated
	public Token readName() {
		return isName()? next() : null;
	}
	
	
	public boolean isNumber() {
		return isEOF()? false : get().kind == TokenKind.Literal;
	}
	
	@Deprecated
	public Token readNumber() {
		return isNumber()? next() : null;
	}
	
	
	
	public boolean isOperator() {
		return isEOF()? false : get().kind == TokenKind.Operator1;
	}
	
	public boolean isOperator(String str) {
		return isOperator() && isMatch(str);
	}
	
	@Deprecated
	public Token readOperator() {
		return isOperator()? next() : null;
	}
	
	
	
	public boolean isReserved() {
		return isEOF()? false : get().kind == TokenKind.Reserved;
	}
	
	public boolean isReserved(String str) {
		return isReserved() && str.equals(get().string);
	}
	
	@Deprecated
	public Token readReserved() {
		return isReserved()? next() : null;
	}
}

File: src/main/scala/marg/token/TokenSet.scala
package marg.token

import java.io.EOFException

import marg.exception.ParseException
import marg.token.TokenKind._

@Deprecated
class TokenSet(ls: List[Token]) extends Iterator[Token] {
  private var offset = 0

  override def next() = {
    checkEOF()
    offset += 1
    ls(offset - 1)
  }

  override def hasNext = !isEOF


  def Offset = offset

  def previous() = if (offset > 0) offset -= 1

  def previous(k: Int) = if (offset >= k) offset -= k


  def get = ls(offset)

  def is(s: String) = s == get.String

  def isEOF = get.isEOF

  def checkEOF() = if (isEOF) throw new EOFException

  def read(args: String*): Boolean = {
    for (t <- args) if (t != next().String) return false
    true
  }

  def readP(message: String, args: String*) =
    for (t <- args) {
      checkEOF()
      if (next().String != t)
        throw new ParseException("Syntax Error: " + message, this)
    }


  def isEOL = !isEOF && (get.Kind eq EOL)

  def isName = !isEOF && (get.Kind eq Identifier)

  def isNumber = !isEOF && (get.Kind eq IntLiteral)

  def isBool = !isEOF && (get.Kind eq BoolLiteral)

  def isOperator = !isEOF && (get.Kind eq Operator)

  def isReserved = !isEOF && (get.Kind eq Reserved)

}
